File: tags/sdk-1.0-before-interface-removal/implementation/UrakawaToolkit/media/data/IMediaData.cs
using System;
using System.Collections.Generic;
using System.Text;
using urakawa.media;

namespace urakawa.media.data
{
	/// <summary>
	/// Interface for a generic <see cref="IMediaData"/>. Uses <see cref="IDataProvider"/>s to store actual data
	/// </summary>
	public interface IMediaData : xuk.IXukAble, IValueEquatable<IMediaData>
	{
		/// <summary>
		/// Gets the <see cref="IMediaDataManager"/> associated with <c>this</c>
		/// </summary>
		/// <returns>The assicoated <see cref="IMediaDataManager"/></returns>
		/// <exception cref="exception.IsNotInitializedException">
		/// Thrown when <c>this</c> has not been associated with a <see cref="IMediaDataManager"/>
		/// </exception>
		IMediaDataManager getMediaDataManager();

		/// <summary>
		/// Associates <c>this</c> with a <see cref="IMediaDataManager"/> - 
		/// initializer that is called in method <see cref="IMediaDataManager.addMediaData"/> method. 
		/// Calling the initializer elsewhere may corrupt the data model.
		/// </summary>
		/// <param name="mngr">The <see cref="IMediaDataManager"/></param>
		/// <exception cref="exception.MethodParameterIsNullException">
		/// Thrown when <paramref name="mngr"/> is <c>null</c>
		/// </exception>
		/// <exception cref="exception.IsAlreadyInitializedException">
		/// Thrown when <c>this</c> has already been associated with a <see cref="IMediaDataManager"/>
		/// </exception>
		void setMediaDataManager(IMediaDataManager mngr);

		/// <summary>
		/// Gets the UID of <c>this</c> within the .
		/// Convenience for <c><see cref="getMediaDataManager"/>().<see cref="IMediaDataManager.getUidOfMediaData"/>(this)</c>
		/// </summary>
		/// <returns>The UID</returns>
		string getUid();
		
		/// <summary>
		/// Gets the name of <c>this</c>
		/// </summary>
		/// <returns></returns>
		string getName();
		
		/// <summary>
		/// Sets the name of <c>this</c>
		/// </summary>
		/// <param name="newName">The new name - must not be <c>null</c></param>
		/// <exception cref="exception.MethodParameterIsNullException">
		/// Thrown when <paramref name="newName"/> is <c>null</c>
		/// </exception>
		void setName(string newName);

		/// <summary>
		/// Deletes <c>this</c>, detaching from the owning manager
		/// </summary>
		void delete();

		/// <summary>
		/// Copies <c>this</c> including any underlying <see cref="IDataProvider"/>s
		/// </summary>
		/// <returns></returns>
		IMediaData copy();

		/// <summary>
		/// Gets a list of the <see cref="IDataProvider"/>s used by <c>this</c>
		/// </summary>
		/// <returns>The list</returns>
		List<IDataProvider> getListOfUsedDataProviders();
	}
}

File: java/core/src/org/daisy/urakawa/media/data/IMediaData.java
package org.daisy.urakawa.media.data;

import java.util.List;

import org.daisy.urakawa.FactoryCannotCreateTypeException;
import org.daisy.urakawa.IPresentation;
import org.daisy.urakawa.IValueEquatable;
import org.daisy.urakawa.IWithPresentation;
import org.daisy.urakawa.event.DataModelChangedEvent;
import org.daisy.urakawa.event.IEventHandler;
import org.daisy.urakawa.exception.MethodParameterIsNullException;
import org.daisy.urakawa.xuk.IXukAble;

/**
 * A IMediaData defines the actual resource where the data for the media object
 * is stored. One or more DataProviders can be associated to this single
 * IMediaData resource. For example, it can be a single file (IFileDataProvider),
 * but it could also be a sequence of various sources (e.g. MemoryDataProvider,
 * MySQLDataProvider, HTTPDataProvider, etc.).
 * 
 * @leafInterface see {@link org.daisy.urakawa.LeafInterface}
 * @see org.daisy.urakawa.LeafInterface
 * @stereotype OptionalLeafInterface
 * @depend - Aggregation 0..n org.daisy.urakawa.media.data.IDataProvider
 * @depend - Clone - org.daisy.urakawa.media.data.IMediaData
 * @depend - Aggregation 1 org.daisy.urakawa.media.data.IMediaDataManager
 * 
 */
public interface IMediaData extends IWithName, IWithPresentation, IXukAble,
		IValueEquatable<IMediaData>, IEventHandler<DataModelChangedEvent> {

	/**
	 * Convenience method to get the UID of this mediadata via
	 * getMediaDataManager().getUidOfMediaData()
	 * 
	 * @return the UID
	 */
	public String getUID();

	/**
	 * Deletes the IMediaData detaching it from it's manager and releasing any
	 * DataProviders used
	 */
	public void delete();

	/**
	 * <p>
	 * Cloning method
	 * </p>
	 * 
	 * @return a copy.
	 */
	public IMediaData copy();

	/**
	 * Returns the DataProviders used by this mediadata
	 * 
	 * @return a non-null list of IDataProvider, potentially empty
	 */
	public List<IDataProvider> getListOfUsedDataProviders();

	/**
	 * @param destPres
	 * @return can return null in case of failure.
	 * @throws FactoryCannotCreateTypeException
	 * @tagvalue Exceptions "FactoryCannotCreateType-MethodParameterIsNull"
	 * @throws MethodParameterIsNullException
	 *             NULL method parameters are forbidden
	 */
	public IMediaData export(IPresentation destPres)
			throws FactoryCannotCreateTypeException,
			MethodParameterIsNullException;
}
